<a href="https://colab.research.google.com/github/saifinuha/data-science-2026/blob/main/Pertemuan12_Saifin_Nuha_240401010257.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Saifin Nuha, 240401010257, IF403

In [1]:
import pandas as pd, numpy as np
import matplotlib.pyplot as plt

np.random.seed(42)
produk = ['Roti', 'Selai', 'Susu', 'Sereal', 'Telur',
          'Keju', 'Kopi', 'Gula', 'Teh', 'Mentega']

# Buat 50 transaksi, tiap transaksi berisi 2-5 produk
transaksi = []
for _ in range(50):
  n_item = np.random.randint(2, 6)
  transaksi.append(list(np.random.choice(produk, n_item, replace=False)))

# Suntikkan pola: Roti sering bersama Selai
for i in range(0, 20):
  if 'Roti' in transaksi[i] and 'Selai' not in transaksi[i]:
    transaksi[i].append('Selai')

print('Contoh transaksi:', transaksi[:3])
print('Jumlah transaksi:', len(transaksi))

Contoh transaksi: [[np.str_('Keju'), np.str_('Roti'), np.str_('Mentega'), np.str_('Kopi'), 'Selai'], [np.str_('Roti'), np.str_('Kopi'), np.str_('Teh'), np.str_('Selai'), np.str_('Mentega')], [np.str_('Kopi'), np.str_('Susu'), np.str_('Teh')]]
Jumlah transaksi: 50


In [2]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(transaksi).transform(transaksi)
df = pd.DataFrame(te_ary, columns=te.columns_)
print(df.head())

    Gula   Keju   Kopi  Mentega   Roti  Selai  Sereal   Susu    Teh  Telur
0  False   True   True     True   True   True   False  False  False  False
1  False  False   True     True   True   True   False  False   True  False
2  False  False   True    False  False  False   False   True   True  False
3  False   True  False    False  False   True   False  False   True   True
4   True   True  False     True  False  False   False   True  False  False


In [3]:
from mlxtend.frequent_patterns import apriori

for ms in [0.05, 0.1, 0.2]:
  freq = apriori(df, min_support=ms, use_colnames=True)
  print(f'min_support={ms}: {len(freq)} itemset ditemukan')

# Gunakan min_support yang menghasilkan jumlah itemset wajar (tidak 0, tidak ratusan)
freq_items = apriori(df, min_support=0.1, use_colnames=True)
freq_items = freq_items.sort_values('support', ascending=False)
print(freq_items.head(10))

min_support=0.05: 74 itemset ditemukan
min_support=0.1: 44 itemset ditemukan
min_support=0.2: 13 itemset ditemukan
    support      itemsets
5      0.52       (Selai)
8      0.46         (Teh)
3      0.42     (Mentega)
9      0.36       (Telur)
1      0.34        (Keju)
0      0.32        (Gula)
2      0.32        (Kopi)
4      0.32        (Roti)
7      0.32        (Susu)
36     0.24  (Selai, Teh)


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [4]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(freq_items, metric='confidence',
                          min_threshold=0.5)
rules = rules[rules['lift'] > 1].sort_values('lift', ascending=False)

print(rules[['antecedents', 'consequents',
            'support', 'confidence', 'lift']].head(10))

         antecedents consequents  support  confidence      lift
10       (Keju, Teh)     (Telur)     0.12    0.857143  2.380952
15  (Selai, Mentega)      (Kopi)     0.10    0.625000  1.953125
12      (Gula, Roti)     (Selai)     0.10    1.000000  1.923077
7           (Sereal)   (Mentega)     0.14    0.777778  1.851852
9       (Teh, Telur)      (Keju)     0.12    0.600000  1.764706
13     (Kopi, Selai)   (Mentega)     0.10    0.714286  1.700680
8      (Keju, Telur)       (Teh)     0.12    0.750000  1.630435
11     (Selai, Gula)      (Roti)     0.10    0.500000  1.562500
14   (Kopi, Mentega)     (Selai)     0.10    0.714286  1.373626
1             (Roti)     (Selai)     0.22    0.687500  1.322115


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## Interpretasi Aturan Asosiasi

### Aturan dengan Lift Tertinggi
Aturan dengan nilai `lift` tertinggi adalah **`(Keju, Teh)` -> `(Telur)`** dengan `lift` sebesar **2.38**. Ini berarti pembeli yang membeli Keju dan Teh secara bersamaan **2.38 kali lebih mungkin** untuk membeli Telur dibandingkan dengan probabilitas membeli Telur secara acak. Tingginya nilai `lift` menunjukkan hubungan yang kuat antara pembelian Keju dan Teh dengan pembelian Telur.

### Apakah Masuk Akal secara Bisnis (mis. Roti -> Selai)?

Ya, beberapa aturan memang masuk akal secara bisnis:

*   **`(Roti)` -> `(Selai)`**: Aturan ini memiliki `lift` **1.32**. Ini sangat masuk akal secara bisnis karena Roti dan Selai adalah produk yang sering dikonsumsi bersama. Tingginya kepercayaan (`confidence`) (68.75%) menunjukkan bahwa sebagian besar pembeli Roti juga membeli Selai. Ini adalah pola yang bisa dimanfaatkan untuk penempatan produk atau promosi bundel.

*   **`(Gula, Roti)` -> `(Selai)`**: Aturan ini memiliki `lift` **1.92**. Ini juga sangat logis karena jika seseorang membeli Roti dan Gula, kemungkinan besar mereka juga akan membeli Selai, mungkin untuk membuat sarapan atau camilan. `Confidence` 100% menunjukkan bahwa setiap kali Gula dan Roti dibeli bersama, Selai juga selalu dibeli.

*   **`(Keju, Teh)` -> `(Telur)`**: Meskipun memiliki `lift` tertinggi, hubungan antara Keju, Teh, dan Telur mungkin tidak sejelas Roti dan Selai. Namun, ini bisa menunjukkan pola pembelian untuk sarapan atau makanan ringan tertentu di mana ketiga item ini sering disertakan. Misalnya, seseorang yang membuat sarapan yang melibatkan telur (mungkin telur orak-arik dengan keju) dan minum teh. Ini mungkin memerlukan analisis lebih lanjut atau pemahaman tentang kebiasaan konsumsi target pasar.

Secara keseluruhan, analisis aturan asosiasi ini memberikan wawasan berharga tentang pola pembelian pelanggan, yang dapat digunakan untuk strategi pemasaran, penempatan produk, atau rekomendasi.

In [6]:
from sklearn.metrics.pairwise import cosine_similarity

katalog = pd.DataFrame({
          'produk': produk,
          'kategori': ['Bakery','Bakery','Dairy','Bakery','Dairy',
                        'Dairy','Minuman','Bumbu','Minuman','Dairy']
          })

fitur = pd.get_dummies(katalog['kategori'])
sim_matrix = cosine_similarity(fitur)

def rekomendasi_serupa(nama_produk, top_n=3):
  idx = katalog.index[katalog['produk'] == nama_produk][0]
  skor = list(enumerate(sim_matrix[idx]))
  skor = sorted(skor, key=lambda x: x[1], reverse=True)
  skor = [s for s in skor if s[0] != idx][:top_n]
  return katalog.iloc[[i for i, _ in skor]]['produk'].tolist()

print('Mirip dengan Roti:', rekomendasi_serupa('Roti'))

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

Mirip dengan Roti: ['Selai', 'Sereal', 'Susu']


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [7]:
produk_target = 'Roti'

# Dari association rules: cari consequents dari aturan yang antecedent-nya mengandung produk_target
rules_terkait = rules[rules['antecedents'].apply(
              lambda x: produk_target in x)]

print('Rekomendasi dari Association Rules:')
print(rules_terkait[['consequents', 'lift']].head())
print('Rekomendasi dari Content-Based:', rekomendasi_serupa(produk_target))

Rekomendasi dari Association Rules:
   consequents      lift
12     (Selai)  1.923077
1      (Selai)  1.322115
Rekomendasi dari Content-Based: ['Selai', 'Sereal', 'Susu']


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

## Diskusi Rekomendasi: Asosiasi vs. Content-Based

### Konsistensi Rekomendasi
Untuk `produk_target = 'Roti'`:

*   **Rekomendasi dari Association Rules:** Output menunjukkan `(Selai)` sebagai konsekuensi dengan `lift` tinggi (1.92 dan 1.32). Ini sangat konsisten dengan pola pembelian yang disuntikkan di awal data (`Roti` sering bersama `Selai`).

*   **Rekomendasi dari Content-Based:** Output menunjukkan `['Selai', 'Sereal', 'Susu']`. Produk-produk ini termasuk dalam kategori yang sama (`Bakery` untuk `Selai`, `Bakery` untuk `Sereal`) atau kategori terkait (`Dairy` untuk `Susu`) dengan `Roti` (`Bakery`).

Terlihat bahwa kedua pendekatan memberikan rekomendasi yang cukup konsisten, terutama dalam mengidentifikasi **Selai** sebagai item yang sangat relevan dengan **Roti**. Pendekatan Content-Based juga memperluas rekomendasi ke produk lain dalam kategori serupa seperti Sereal dan Susu.

### Kapan Menggunakan Setiap Pendekatan atau Menggabungkan Keduanya?

#### 1. Association Rules
*   **Kapan Digunakan:** Ideal untuk menemukan produk komplementer (sering dibeli bersama) dan mengidentifikasi pola pembelian lintas kategori yang tidak selalu terlihat dari atribut produk saja. Sangat berguna untuk strategi _bundling_, penempatan produk di toko (misalnya, menempatkan Selai di dekat Roti), dan rekomendasi "orang yang membeli ini juga membeli itu".
*   **Kekuatan:** Menangkap hubungan perilaku pelanggan secara langsung dari data transaksi.
*   **Kelemahan:** Memerlukan data transaksi yang cukup besar dan bisa menghasilkan terlalu banyak aturan jika `min_support` terlalu rendah, atau kehilangan aturan penting jika terlalu tinggi. Sulit merekomendasikan produk baru karena belum memiliki riwayat transaksi.

#### 2. Content-Based Filtering
*   **Kapan Digunakan:** Cocok untuk merekomendasikan produk serupa berdasarkan atribut intrinsik produk (kategori, merk, fitur, dll.). Berguna ketika ingin menyarankan alternatif atau variasi dari produk yang diminati pelanggan. Sangat baik untuk mengatasi masalah _cold start_ (produk baru dapat direkomendasikan jika atributnya cocok).
*   **Kekuatan:** Tidak memerlukan data riwayat interaksi pengguna-item; hanya bergantung pada deskripsi produk. Bekerja dengan baik untuk produk baru.
*   **Kelemahan:** Terbatas pada merekomendasikan item yang `serupa` dalam atribut, mungkin melewatkan hubungan komplementer yang tidak terkait atribut langsung (misalnya, `popok` dan `bir` dari contoh klasik).

#### 3. Pendekatan Hybrid (Menggabungkan Keduanya)
*   **Kapan Digunakan:** Dalam banyak kasus di dunia nyata, pendekatan hybrid adalah yang paling efektif. Menggabungkan kedua metode dapat memanfaatkan kekuatan masing-masing dan menutupi kelemahannya.
    *   Misalnya, setelah pelanggan membeli **Roti**:
        *   **Association Rules** dapat merekomendasikan **Selai** (karena sering dibeli bersama).
        *   **Content-Based** dapat merekomendasikan **Sereal** atau jenis roti lainnya (karena mereka dari kategori yang sama atau serupa).
*   **Manfaat:** Memberikan rekomendasi yang lebih kaya dan relevan. Mampu menangani _cold start_ untuk produk dan juga menemukan pola pembelian yang kompleks.
*   **Implementasi:** Bisa dilakukan dengan berbagai cara, misalnya dengan menggabungkan skor dari kedua sistem, atau menggunakan satu sistem untuk menyaring kandidat dan sistem lain untuk memberi peringkat akhir.